sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfigurationprovider/_azureappconfigurationprovider.py

```diff
@@ -0,0 +1,183 @@
+                    # Deals with possible null value via Rest API
+                    provider._dict[provider.__trim(config.key)] = config.value
+                elif config.content_type == "application/vnd.microsoft.appconfig.keyvaultref+json;charset=utf-8":
+                    provider.__resolve_keyvault_references(
+                        config, key_vault_options, secret_clients)
+                elif "application/json" in config.content_type:
+                    j_object = json.loads(config.value)
+                    provider._dict[provider.__trim(config.key)] = j_object
+                else:
+                    provider._dict[provider.__trim(config.key)] = config.value
+        return provider
+
+    def __buildprovider(self, connection_string, endpoint, credential, key_vault_options):
+        headers = {}
+        correlation_context = "RequestType=Startup"
+
+        if (key_vault_options is not None and
+                (key_vault_options.credential is not None or key_vault_options.secret_clients is not None or
+                 key_vault_options.secret_resolver is not None)):
+            correlation_context += ",UsesKeyVault"
+
+        headers["Correlation-Context"] = correlation_context
+        useragent = USER_AGENT
+
+        if connection_string is not None:
+            self._client = AzureAppConfigurationClient.from_connection_string(
+                connection_string, user_agent=useragent, headers=headers)
+            return
+        self._client = AzureAppConfigurationClient(
+            endpoint, credential, user_agent=useragent, headers=headers)
+
+    def __resolve_keyvault_references(self, config, key_vault_options, secret_clients):
+        if key_vault_options is None:
+            raise AttributeError(
+                "Key Vault options must be set to resolve Key Vault references.")
+        j_object = json.loads(config.value)
+        uri_value = j_object['uri']
+
+        uri = urlparse(uri_value)
+
+        key_vault_uri = "https://" + uri.hostname
+        key_vault_secret_prefix = "/secrets/"
+        key_vault_secret_name = uri.path[len(
+            key_vault_secret_prefix):]
+        if key_vault_options.credential is not None:
+            secret_client = None
+
+            # Clients only should be made once, will reuse if client already made
+            for client_uri in secret_clients:
+                if client_uri == key_vault_uri:
+                    secret_client = secret_clients[client_uri]
+                    break
+            if secret_client is None:
+                secret_client = SecretClient(
+                    vault_url=key_vault_uri, credential=key_vault_options.credential)
+                secret_clients[key_vault_uri] = secret_client
+                secret = secret_client.get_secret(
+                    key_vault_secret_name)
+                self._dict[self.__trim(config.key)] = secret.value
+                return
+            raise AttributeError(
```

Comment:
We are inside the if block that says `if key_vault_options.credential is not None`.
At this point, if we dont have any secret client, we should still try to look for secret_resolver before raising AttributeError

Reply:
This `AttributeError` is for the key vault reference not having a uri field. It doesn't matter if a client exists or not yet.

Reply:
The error message is misleading in that case.

Reply:
This should be fixed with the rework.